### PR TITLE
chore: specify fees in invitations and if present, charge those fees

### LINKED ISSUE #3689: Contracts can attach fees and `expiry` timestamps to invitations. Users are charged the fee.
Invitation details will have two new (optional) properties: `fee` (a RUN amount) and `expiry` (a timestamp).

These are specified by the contract code in relative terms (fee is HIGH or LOW, expiry is SHORT or LONG), and translated to absolute terms by Zoe. 

Zoe charges the fee associated with an invitation when the user uses the invitation to make an offer. Unlike the other fees that Zoe charges which go into a general collection purse, this fee is deposited directly into the `feePurse` for the contract creator. This is so that the contract creator can charge users fees to cover the contract execution costs. 

Thus, `E(zoe).offer` will result in two fees being charged: the `offerFee` determined in the Zoe configuration, and the fee (if any) attached to the invitation. 

As part of the PR which closes this issue, we should add fees to the AMM swap and Treasury open vault invitations. 
As part of the PR which closes this issue, we should allow users (and their wallet) to get the ZoeFeeConfig from Zoe. This will mean adding another method to Zoe. Something like "getConfiguration".

### PR DESCRIPTION
**Pull request by katelynsills**
Closes #3689 

Stacked on #3688

### CODE REVIEW

File: packages/zoe/src/zoeService/makeInvitation.js [hunk 1]
```diff
@@ -20,31 +29,41 @@ export const createInvitationKit = shutdownZoeVat => {
...
         'string',
         X`The description ${description} must be a string`,
       );
+      const absoluteFee = translateFee(relativeFee);
+      const absoluteExpiry = await translateExpiry(relativeExpiry);
+
+      let feeInfo = {};
+      if (absoluteFee && absoluteExpiry) {
+        feeInfo = { fee: absoluteFee, expiry: absoluteExpiry };
+      }
```
Comment (erights) [inline packages/zoe/src/zoeService/makeInvitation.js]: (This one is really important so making visible prior to review.) 

As of this PR, Fred will take the `fee` and `expiry` properties in the invite amount to be authentic. At least to have been translated by Zoe's internal `translateFee` and `translateExpiry` and have whatever invariants follow from that. However, by allowing the `feeInfo` to be empty, it would fail to override bad `fee` and `expiry` properties provided by the contract in the `customProperties`.

```suggestion
      const feeInfo = { fee: absoluteFee, expiry: absoluteExpiry };
```

Comment (katelynsills) [reply packages/zoe/src/zoeService/makeInvitation.js]: good catch!

File: packages/zoe/src/zoeService/makeInvitation.js [hunk 2]
```diff
@@ -2,7 +2,16 @@
...
-export const createInvitationKit = shutdownZoeVat => {
+/**
+ * @param {ShutdownWithFailure | undefined} shutdownZoeVat
+ * @param {TranslateFee | (() => void)} translateFee
+ * @param {TranslateExpiry | (() => void)} translateExpiry
+ */
+export const createInvitationKit = (
+  shutdownZoeVat = undefined,
+  translateFee = () => {},
+  translateExpiry = () => {},
```
Comment (erights) [inline packages/zoe/src/zoeService/makeInvitation.js]: Since you're using the results of these functions, if you really want them to default to returning `undefined`, which is their current behavior, better to make this more explicit.

```suggestion
  translateFee = () => undefined,
  translateExpiry = () => undefined,
```

But is returning `undefined` in fact the default behavior you really want?

File: packages/zoe/src/zoeService/makeInvitation.js [hunk 3]
```diff
@@ -20,31 +29,41 @@ export const createInvitationKit = shutdownZoeVat => {
...
+      // instance for the contract, even if customProperties includes
+      // a property called `instance`.
       const invitationAmount = AmountMath.make(invitationKit.brand, [
         {
           ...customProperties,
           description,
           handle: invitationHandle,
           instance,
           installation,
+          ...feeInfo, // may be empty
```
Comment (erights) [inline packages/zoe/src/zoeService/makeInvitation.js]: I followed the translateExpiry and see it returns a `Timestamp` which is a raw bigint. Which is meaningless to Fred without knowing what `timeAuthority` it is relative to. That's why Fred needs it in the coveredCall.

Since you already define an explicit `Timestamp` type, that currently aliases to `bigint`, would redefining it be a good place to take @dtribble 's suggestion of passing around a number-timeAuthority pair, like we do amounts? Whether we do it in the type of not, consider passing a pair record here rather than just adding `timeAuthority` to the authoritative property names.

File: packages/zoe/src/zoeService/makeInvitation.js (hunk 1 above)
Comment (katelynsills) [reply packages/zoe/src/zoeService/makeInvitation.js]: Fixed

File: packages/zoe/src/zoeService/offer/offer.js [hunk 4]
```diff
@@ -36,17 +39,33 @@ export const makeOffer = (
...
+    } = await burnInvitation(invitationIssuer, invitation);
     // AWAIT ///
 
-    await chargeZoeFee(feePurse, offerFeeAmount);
-
     const instanceAdmin = getInstanceAdmin(instanceHandle);
     instanceAdmin.assertAcceptingOffers();
 
+    if (timeAuthority && expiry && fee) {
+      // TODO: is there a way to make this a top-level await?
```
Comment (erights) [inline packages/zoe/src/zoeService/offer/offer.js:58]: Ok I tried and failed to rewrite without significant loss in readability. Thus, no change suggested for this PR. But let's remember to come back to this one.

File: packages/zoe/src/zoeService/zoe.js [hunk 5]
```diff
@@ -181,12 +222,20 @@ const makeZoeKit = (
...
+    });
+  };
+
   /** @type {ZoeServiceFeePurseRequired} */
   const zoeService = Far('zoeServiceFeePurseRequired', {
     install,
     startInstance,
     offer,
-    makeFeePurse,
+    makeFeePurse: async () => makeFeePurse(),
```
Comment (erights) [inline packages/zoe/src/zoeService/zoe.js:239]: No objection, but curious why this change?

**Review by erights — APPROVED** [review]
LGTM

File: packages/zoe/src/zoeService/makeInvitation.js (hunk 3 above)
Comment (katelynsills) [reply packages/zoe/src/zoeService/makeInvitation.js]: Ah, good point. My idea here was that Fred could call E(zoe).getConfiguration() to get the Zoe timeAuthority, so there would be no confusion about what timeAuthority was being used. I think that takes care of any Fred problem. When we rethink how metering works in general, we should revisit this to see if it can be easily improved.

File: packages/zoe/src/zoeService/zoe.js (hunk 5 above)
Comment (katelynsills) [reply packages/zoe/src/zoeService/zoe.js:239]: I needed a version of `makeFeePurse` that returned a promise (the external version) and a version that was synchronous. This is converting to the external version.

File: packages/zoe/src/zoeService/makeInvitation.js (hunk 2 above)
Comment (katelynsills) [reply packages/zoe/src/zoeService/makeInvitation.js]: Hmm, what do you think would be preferable?

File: packages/zoe/src/zoeService/zoe.js (hunk 5 above)
Comment (erights) [reply packages/zoe/src/zoeService/zoe.js:239]: Good. Resolving.

File: packages/zoe/src/zoeService/makeInvitation.js (hunk 3 above)
Comment (katelynsills) [reply packages/zoe/src/zoeService/makeInvitation.js]: I'll add a new property `zoeTimeAuthority` to the invitation details to solve this, like we discussed.

File: packages/zoe/src/zoeService/makeInvitation.js (hunk 2 above)
Comment (erights) [reply packages/zoe/src/zoeService/makeInvitation.js]: I just looked at where they're used, and we have a worse problem. That same problematic `if` says
```js
if (timeAuthority && expiry && fee) {
```
which interprets falsy as ignore. But `expiry` is absolute time according to some time authority, and `0` is falsy. Even if it could never happen with the only possible time authority, this code should not be written to assume that `0` is not an absolute time that might matter.

Given that, I would keep the `undefined` from my suggestion above (still to be made explicit). But please revise that `if` to check `=== undefined` explicitly.

Comment (katelynsills) [reply packages/zoe/src/zoeService/makeInvitation.js]: Done
